src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Allows limited reuse of heap buffers to improve memory pressure. This cache does not ensure
+    /// that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
+    {
+        private readonly ulong _minSize;
+        private readonly ulong _maxSize;
+
+        private LowLevelConcurrentStack<SafeHeapHandle> handleCache = new LowLevelConcurrentStack<SafeHeapHandle>();
```

Comment:
One of the biggest advantages here is that we don't have to keep reallocating buffer space. It is absolutely key to my design. Taking that out really has a big negative impact on the perf I measured. Keeping a buffer alive for every thread doesn't make sense- even at two legacy max paths that is 1MB. For a "typical" long path that is easily 2-4X that.

In addition I didn't see any advantage of keeping a thread local StringBuffer- it _negatively_ impacted the measurements I took. With such evidence I think it makes more sense to add it in once we've found a real measurement that shows how it improves a specific scenario.

> We're using native memory rather than managed memory in that other type, in order to avoid pinning costs as part of marshaling.

Pinning is just one characteristic. More important is that the memory isn't cleared and that it has a fast copy on realloc.

> We're jumping to P/Invokes for the memory allocation rather than .NET's built-in functions for native memory allocation.

Again, this all about optimizing scratch space. We don't need to zero, we need to be able to grow quickly.

I'll call it out again that I've _completely_ rewritten the logic for Path.GetFullPath. It is fair to look at some of the old patterns for informing changes, but the old patterns won't necessarily play out the same. I took the current behavior and dug into how to make it better using the original patterns to inform my attack on this. I'm not arbitrarily making changes- everything has many hours of investigation behind it.